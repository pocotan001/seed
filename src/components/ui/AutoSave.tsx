import { isEqual } from "lodash";
import * as React from "react";
import { FormSpy, FormSpyRenderProps } from "react-final-form";
import { debounce } from "~/infra/decorators";

interface IAdditionalProps {
  onRequestSave: (values: any) => void | Promise<void>;
}

interface IAutoSaveProps extends FormSpyRenderProps, IAdditionalProps {}

interface IAutoSaveState {
  values: { [name: string]: string };
  isSaved: boolean;
}

/**
 * Automatically submit form values after a debounce period
 *
 * @example
 * const save = values => { ... };
 *
 * <Form>
 *   <AutoSave onRequestSave={save} />
 *   ...
 * </Form>
 */
class AutoSave extends React.PureComponent<IAutoSaveProps, IAutoSaveState> {
  static getDerivedStateFromProps(
    nextProps: IAutoSaveProps,
    prevState: IAutoSaveState
  ): Partial<IAutoSaveState> | null {
    const { values } = prevState;

    if (!isEqual(values, nextProps.values)) {
      return {
        values: nextProps.values,
        isSaved: false
      };
    }

    return null;
  }

  state = {
    values: this.props.values,
    isSaved: false
  };

  componentDidUpdate(_: IAutoSaveProps, prevState: IAutoSaveState) {
    const { values } = this.state;

    if (prevState.values !== values) {
      this.save();
    }
  }

  componentWillUnmount() {
    (this.save as any).cancel();
  }

  @debounce(1000)
  async save() {
    const { onRequestSave } = this.props;

    await onRequestSave(this.state.values);
    this.setState({ isSaved: true });
  }

  render() {
    // Doesn't have to render anything, but it can render saved state
    // `return this.state.isSaved && <p>Saved</p>`
    return null;
  }
}

const AdaptedAutoSave: React.SFC<IAdditionalProps> = props => (
  <FormSpy
    {...props}
    subscription={{ values: true }}
    component={AutoSave as any}
  />
);

export default AdaptedAutoSave;
