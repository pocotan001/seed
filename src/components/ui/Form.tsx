import { FORM_ERROR, FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

export type IFormOnSubmit = (
  values: IFormValues,
  form: FormApi
) => Promise<void | string> | void | string;

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: IFormOnSubmit;
      }
    > {}

const focusOnErrors = createDecorator();

export default class Form extends React.PureComponent<IFormProps> {
  handleSubmit = async (values: IFormValues, form: FormApi): Promise<any> => {
    const { onSubmit } = this.props;
    const err = await onSubmit(values, form);

    if (err) {
      return { [FORM_ERROR]: err };
    }
  };

  render() {
    const { onSubmit, ...rest } = this.props;

    return (
      <FinalForm
        decorators={[focusOnErrors]}
        onSubmit={this.handleSubmit}
        {...rest}
      />
    );
  }
}
