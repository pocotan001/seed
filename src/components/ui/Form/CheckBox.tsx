import * as React from "react";
import margin, { IMarginProps } from "~/components/styles/extends/margin";
import { ITheme } from "~/components/styles/theme";
import styled from "~/components/styles/themedStyledComponents";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";

type IInputAttributes = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "type"
>;

interface IAdditionalProps extends IMarginProps {
  children?: React.ReactNode;
  theme?: ITheme;
}

const CheckBox: React.SFC<
  IFieldRenderProps<IInputAttributes> & IAdditionalProps
> = ({ input, meta, className, children, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return (
    <label className={className}>
      <input {...input} {...rest} aria-invalid={isInvalid} />
      {children && <span className="label">{children}</span>}
    </label>
  );
};

const StyledCheckBox = styled(CheckBox)`
  display: inline-block;
  vertical-align: middle;
  ${margin};

  > input {
    vertical-align: middle;
    margin-right: 8px;
  }

  > .label {
    vertical-align: middle;
  }
`;

const AdaptedCheckBox: React.SFC<
  IFieldProps<IInputAttributes> & IAdditionalProps
> = props => <Field {...props} type="checkbox" component={StyledCheckBox} />;

export default AdaptedCheckBox;
