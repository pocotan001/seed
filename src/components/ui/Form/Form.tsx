import { FormApi } from "final-form";
import createDecorator from "final-form-focus";
import * as React from "react";
import { Form as FinalForm, FormProps } from "react-final-form";

type IFormValues = any;

interface ISubmitErrors {
  [name: string]: string;
}

interface IFormProps
  extends Overwrite<
      FormProps,
      {
        initialValues?: IFormValues;
        onSubmit: (
          values: IFormValues,
          form: FormApi,
          callback?: (errors?: ISubmitErrors) => void
        ) => Promise<ISubmitErrors | void> | ISubmitErrors | void;
      }
    > {}

const focusOnErrors = createDecorator();
const Form: React.SFC<IFormProps> = props => (
  <FinalForm decorators={[focusOnErrors]} {...props} />
);

export default Form;
