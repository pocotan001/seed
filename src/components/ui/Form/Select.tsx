import * as React from "react";
import { ITheme } from "~/components/styles/theme";
import styled from "~/components/styles/themedStyledComponents";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";
import { ITextFieldStyleProps, textFieldStyles } from "./TextField";

type ISelectAttributes = React.SelectHTMLAttributes<HTMLSelectElement>;

interface IAdditionalProps extends ITextFieldStyleProps {
  theme?: ITheme;
}

const Select: React.SFC<
  IFieldRenderProps<ISelectAttributes> & IAdditionalProps
> = ({ input, meta, children, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  if (rest.multiple) {
    input.value = input.value || [];
    rest.size = Array.isArray(children) ? children.length : 1;
  }

  return (
    <select {...input} {...rest} aria-invalid={isInvalid}>
      {children}
    </select>
  );
};

const StyledSelect = styled(Select)`
  ${textFieldStyles};

  &:not([multiple]):not([size]) {
    /* (font-size * line-height + padding) + border-width */
    height: calc(2.3em + 2px);
  }
`;

const AdaptedSelect: React.SFC<
  IFieldProps<ISelectAttributes> & IAdditionalProps
> = props => <Field {...props} type="select" component={StyledSelect} />;

export default AdaptedSelect;
