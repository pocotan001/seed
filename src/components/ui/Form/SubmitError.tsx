import { FORM_ERROR } from "final-form";
import * as React from "react";
import { Field, FieldRenderProps } from "react-final-form";
import { IMarginProps } from "~/components/styles/extends/margin";
import Paragraph from "../Paragraph";

type ISubmitErrorProps = IMarginProps;

const render = ({
  meta: { submitError },
  m,
  mt,
  mr,
  mb,
  ml
}: FieldRenderProps & IMarginProps) =>
  submitError ? (
    <Paragraph
      fz={14}
      c="pink500"
      m={m}
      mt={mt}
      mr={mr}
      mb={mb}
      ml={ml}
      role="alert"
    >
      {submitError}
    </Paragraph>
  ) : null;

/**
 * The whole-form submission error returned by `onSubmit` under the `FORM_ERROR` key
 *
 * @example
 * const handleSubmit = () => ({ [FORM_ERROR]: "Failed" });
 *
 * <Form onSubmit={handleSubmit}>
 *   <SubmitError />
 * </Form>
 */
const SubmitError: React.SFC<ISubmitErrorProps> = props => (
  <Field
    {...props}
    name={FORM_ERROR}
    subscription={{ submitError: true }}
    render={render}
  />
);

export default SubmitError;
