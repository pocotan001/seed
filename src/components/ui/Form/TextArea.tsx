import * as React from "react";
import styled from "styled-components";
import { Field, IFieldProps, IFieldRenderProps } from "./Field";
import { ITextFieldStyleProps, textFieldStyles } from "./TextField";

type ITextAreaAttributes = React.TextareaHTMLAttributes<HTMLTextAreaElement>;
type IAdditionalProps = ITextFieldStyleProps;

const TextArea: React.SFC<
  IFieldRenderProps<ITextAreaAttributes> & IAdditionalProps
> = ({ input, meta, m, mt, mr, mb, ml, ...rest }) => {
  const isInvalid = Boolean(meta.touched && meta.error);

  return <textarea {...input} {...rest} aria-invalid={isInvalid} />;
};

const StyledTextArea = styled(TextArea)`
  resize: vertical;
  ${textFieldStyles};
`;

const AdaptedTextArea: React.SFC<
  IFieldProps<ITextAreaAttributes> & IAdditionalProps
> = props => <Field {...props} component={StyledTextArea} />;

export default AdaptedTextArea;
