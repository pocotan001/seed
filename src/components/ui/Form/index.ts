import AutoSave from "./AutoSave";
import CheckBox from "./CheckBox";
import _Form from "./Form";
import Radio from "./Radio";
import Select from "./Select";
import SubmitError from "./SubmitError";
import TextArea from "./TextArea";
import TextField from "./TextField";
import ValidationError from "./ValidationError";

const Form: typeof _Form & {
  AutoSave: typeof AutoSave;
  CheckBox: typeof CheckBox;
  Radio: typeof Radio;
  Select: typeof Select;
  SubmitError: typeof SubmitError;
  TextArea: typeof TextArea;
  TextField: typeof TextField;
  ValidationError: typeof ValidationError;
} = _Form as any;

Form.AutoSave = AutoSave;
Form.CheckBox = CheckBox;
Form.Radio = Radio;
Form.Select = Select;
Form.SubmitError = SubmitError;
Form.TextArea = TextArea;
Form.TextField = TextField;
Form.ValidationError = ValidationError;

export default Form;
