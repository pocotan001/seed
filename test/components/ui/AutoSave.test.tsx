import * as React from "react";
import * as renderer from "react-test-renderer";
import AutoSave from "~/components/ui/AutoSave";
import Form from "~/components/ui/Form";

const handleSubmit = () => undefined;
const save = () => undefined;

describe("<AutoSave>", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <Form onSubmit={handleSubmit}>
          {() => <AutoSave onRequestSave={save} />}
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
