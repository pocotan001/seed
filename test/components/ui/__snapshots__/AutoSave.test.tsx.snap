// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoSave> should render correctly 1`] = `null`;
